File: pkg/nuclide-debugger-python-rpc/debugger/debugger.py
# Copyright (c) 2015-present, Facebook, Inc.
# All rights reserved.
#
# This source code is licensed under the license found in the LICENSE file in
# the root directory of this source tree.

import bdb
import json
import socket
import sys

METHOD_CONNECT = 'connect'
METHOD_CONTINUE = 'continue'
METHOD_EXIT = 'exit'
METHOD_INIT = 'init'
METHOD_JUMP = 'jump'
METHOD_NEXT = 'next'
METHOD_QUIT = 'quit'
METHOD_RETURN = 'return'
METHOD_START = 'start'
METHOD_STEP = 'step'
METHOD_STOP = 'stop'


def debug(path_to_socket):
    '''Creates a new Debugger that will publish events to path_to_socket.

    Note that this debugger is only designed for single-threaded Python code
    right now. Check out threading.set_trace() and
    https://github.com/fabioz/PyDev.Debugger for a multi-threaded debugger.

    Args:
        path_to_socket (string): absolute path to .sock file
    Returns:
        Debugger once debugging has finished.
    '''
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    io = SocketIO(sock)
    sock.connect(path_to_socket)
    io.send({'method': METHOD_CONNECT})
    init_message = io.next()
    if init_message['method'] != METHOD_INIT:
        raise Exception('Received method other than init: ' + str(init_message))

    # We wait until we have the initial breakpoints before proceeding.
    debugger = Debugger(sock, io)
    for bp in init_message['breakpoints']:
        debugger.set_break(bp['file'], bp['line'])
    cmd = 'execfile(%r)' % sys.argv[0]
    debugger.run(cmd)
    return debugger


class SocketIO(bdb.Bdb):
    '''Reads and writes newline-delimited-JSON via a socket using blocking I/O.
    '''
    def __init__(self, sock):
        self._sock = sock
        self._data = ''

    def send(self, payload):
        '''
        Args:
            payload: Some sort of JSON-serializable object.
        '''
        self._sock.send(json.dumps(payload) + '\n')

    def next(self):
        '''Returns the next parsed JSON value.

        WARNING: This method is blocking.
        '''
        while True:
            index = self._data.find('\n')
            if index != -1:
                data = self._data[:index]
                out = json.loads(data)
                self._data = self._data[index + 1:]
                return out
            else:
                self._data += self._sock.recv(4096)


class Debugger(bdb.Bdb):
    def __init__(self, sock, io):
        '''Assumes sys.argv is in the proper state.
        '''
        bdb.Bdb.__init__(self)
        self._sock = sock
        self._io = io

    def run(self, cmd, globals=None, locals=None):
        bdb.Bdb.run(self, cmd, globals, locals)
        self._send_message({'method': METHOD_EXIT})
        self._sock.close()

    def user_line(self, frame):
        '''Override abstract method.

        One of self.stop_here(frame) or self.break_here(frame) is True.
        '''
        # If we are at a breakpoint, notify the client and block until we
        # are ready to yield control back to the running program.
        filename = frame.f_code.co_filename
        line = frame.f_lineno
        if filename == '<string>' and line == 1:
            # This is a special call that happens when the debugger starts that
            # gives clients a chance to perform any setup before continuing.
            self._send_message({
                'method': METHOD_START,
            })
        else:
            self._send_message({
                'method': METHOD_STOP,
                'file': filename,
                'line': line,
            })
        self._process_events_until_resume(frame, traceback=None)

    def _process_events_until_resume(self, frame, traceback):
        self.stack, self.curindex = self.get_stack(frame, traceback)
        self.curframe = self.stack[self.curindex][0]
        while True:
            message = self._io.next()
            method = message['method']
            if method == METHOD_CONTINUE:
                self.set_continue()
                return
            elif method == METHOD_NEXT:
                self.set_next(self.curframe)
                return
            elif method == METHOD_STEP:
                self.set_step()
                return
            elif method == METHOD_RETURN:
                self.set_return(self.curframe)
                return
            elif method == METHOD_JUMP:
                if self.curindex + 1 != len(self.stack):
                    # You can only jump within the bottom frame.
                    return
                line = message['line']
                self.curframe.f_lineno = line
                self.stack[self.curindex] = self.stack[self.curindex][0], line
                # self.print_stack_entry(self.stack[self.curindex])
                return
            elif method == METHOD_QUIT:
                self.set_quit()
                return

            # User may be trying to do something while the debugger is
            # suspended, such as setting or removing a breakpoint.
            # TODO(mbolin): Handle other events.
            pass

    def _send_message(self, payload):
        self._io.send(payload)

File: pkg/nuclide-debugger-python-rpc/debugger/debugger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.launchDebugger = launchDebugger;

var _rxjsBundlesRxMinJs = require('rxjs/bundles/Rx.min.js');

var _child_process = _interopRequireDefault(require('child_process'));

var _net = _interopRequireDefault(require('net'));

var _nuclideUri;

function _load_nuclideUri() {
  return _nuclideUri = _interopRequireDefault(require('../../commons-node/nuclideUri'));
}

var _split;

function _load_split() {
  return _split = _interopRequireDefault(require('split'));
}

var _uuid;

function _load_uuid() {
  return _uuid = _interopRequireDefault(require('uuid'));
}

var _os = _interopRequireDefault(require('os'));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 *
 * 
 */

const METHOD_CONNECT = 'connect';
const METHOD_EXIT = 'exit';
const METHOD_INIT = 'init';
const METHOD_START = 'start';
const METHOD_STOP = 'stop';

const PARAM_BREAKPOINTS = 'breakpoints';
const PARAM_METHOD = 'method';

function launchDebugger(commander, initialBreakpoints, pathToPythonExecutable, pythonArgs) {
  return _rxjsBundlesRxMinJs.Observable.create(observer => {
    function log(message) {
      observer.next({ event: 'log', message });
    }

    const server = _net.default.createServer(connection => {
      // For simplicity, we use newline-delimited-JSON as our wire protocol.
      function write(dict) {
        connection.write(JSON.stringify(dict) + '\n');
      }

      // Listen to events broadcast from the Python debugger.
      connection.pipe((0, (_split || _load_split()).default)(JSON.parse, /* mapper */null, { trailing: false })).on('data', args => {
        const method = args[PARAM_METHOD];
        if (method === METHOD_CONNECT) {
          // On initial connection, we should send the breakpoints over.
          write({ [PARAM_METHOD]: METHOD_INIT, [PARAM_BREAKPOINTS]: initialBreakpoints });
          observer.next({ event: 'connected' });
        } else if (method === METHOD_STOP) {
          const { file, line } = args;
          observer.next({ event: 'stop', file, line });
        } else if (method === METHOD_EXIT) {
          observer.next({ event: 'exit' });
          connection.end();
        } else if (method === METHOD_START) {
          observer.next({ event: 'start' });
        } else {
          const error = new Error(`Unrecognized message: ${JSON.stringify(args)}`);
          observer.error(error);
        }
      });

      // Take requests from the input commander and pass them through to the Python debugger.
      // TODO(mbolin): If a `quit` message comes in, we should tear down everything from here
      // because the Python code may be locked up such that it won't get the message.
      commander.subscribe(write, error => log(`Unexpected error from commander: ${String(error)}`), () => log('Apparently the commander is done.'));

      connection.on('end', () => {
        // In the current design, we only expect there to be one connection ever, so when it
        // disconnects, we can shut down the server.
        server.close();
        observer.complete();
      });
    });

    server.on('error', err => {
      observer.error(err);
      throw err;
    });

    const socketPath = createSocketPath();
    server.listen({ path: socketPath }, () => {
      log(`listening for connections on ${socketPath}. About to run python.`);

      // The connection is set up, so now we can launch our Python program.
      const pythonDebugger = (_nuclideUri || _load_nuclideUri()).default.join(__dirname, 'main.py');
      const args = [pythonDebugger, socketPath].concat(pythonArgs);
      const python = _child_process.default.spawn(pathToPythonExecutable, args);

      /* eslint-disable no-console */
      // TODO(mbolin): These do not seem to be fired until the debugger finishes.
      // Probably need to handle things differently in debugger.py.
      python.stdout.on('data', data => console.log(`python stdout: ${data}`));
      python.stderr.on('data', data => console.log(`python stderr: ${data}`));
      /* eslint-enable no-console */
    });
  });
}

function createSocketPath() {
  return (_nuclideUri || _load_nuclideUri()).default.join(_os.default.tmpdir(), `${(_uuid || _load_uuid()).default.v4()}.sock`);
}